docs/configure.md
# config.json (file)
## File example
```
{
  "HttpServer": {
    "Listen":    ":8080",
    "EnableTLS": false,
    "CertFile":  "",
    "KeyFile":   ""
  },
  "Logging": {
  	"Level": "Info",
  	"LogRequest": true,
  	"Log404Request": true
  },
  "ShowHiddenFiles": false,
  "HumanFileSize": true,
  "ModTimeFormat": "2006 Jan 2 15:04"
}
```

## HTTP server
### Listen
Specifies the **port and ip** address that the server is listening to (default: `":8080"`).

```
{
  "HttpServer": {
    "Listen": ":8080"
  }
}

```

### Enable TLS
`EnableTLS` - enables TLS encryption (default: `false`)

`CertFile` - specify the path to the certificate file (default: `""`)

`KeyFile` - specify the path to the key file (default: `""`)

```
{
  "HttpServer": {
    "EnableTLS": true,
    "CertFile": "/etc/ssl/cert",
    "KeyFile": "/etc/ssl/key"
  }
}

```

## Logging
`Level` - specifies the events to be logged. Possible values: `Info`, `Warning`, `Error` (default: `"Info"`)
`LogRequest` - logs all successful requests (default: `true`)
`Log404Request` - logs all 404 error requests (default: `true`)

```
{
  "Logging": {
  	"Level": "Info",
  	"LogRequest": true,
  	"Log404Request": true
  }
}
```

## Web Interface
### Show hidden files
Does not display files that begin with a dot. If the file or directory name is entered correctly, it can still be used (default: `false`).

```
{
  "ShowHiddenFiles": false
}

```

### Human readable file size
Displays file sizes in human-readable format. For example: `4096=4K`, `5200000=5M` (default: `true`).

```
{
  "HumanFileSize": true
}

```

### Modification time format
Sets the custom date and time format of the modification. Learn more about the date and time format in GO: [src/time/format.go](https://golang.org/src/time/format.go) (default: `"2006 Jan 2 15:04"`).

```
{
  "ModTimeFormat": "2006 Jan 2 15:04"
}

```

# index.tmpl (file)
If you want to write your own HTML page template, use the [GO documentation](https://pkg.go.dev/html/template). The example below shows the template example.
## Template example
```
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>{{.Path}} - Horizon</title>
  </head>
  <body>
    <h2>{{.Path}}</h2>
    <p><a href={{.UpPath}}>Go top</a></p>
    <table align='left' border=1 cellpadding=5>
      <th align='left'>Name</th>
      <th align='left'>Size (bit)</th>
      <th align='left'>Mode</th>
      <th align='left'>Modification time</th>
      <th align='left'>Owner</th>
      <th align='left'>Group</th>
      {{range .Files}}
        <tr>
          <td><a href={{.Path}}>{{.Name}}</a></td>
          <td>{{.Size}}</td>
          <td>{{.Mode}}</td>
          <td>{{.ModTime}}</td>
          <td>{{.Owner}} ({{.Uid}})</td>
          <td>{{.Group}} ({{.Gid}})</td>
        </tr>
      {{end}}
    </table>
  </body>
</html>
```

## Default templates
The templates used by default are in the source code:
- UNIX-like: `./internal/settings/settings_unix.go`
- Windows: `./internal/settings/settings_windows.go`

# resources (dir)
Contains custom resources that will be available but will not appear in the file list. For example it can be used for CSS styles or favicon.ico.


internal/settings/settings.go
/*
   Copyright 2021 Leonid Maslakov

   License: GPL-3.0-or-later

   This file is part of Horizon.

   Horizon is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   Horizon is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with Horizon.  If not, see <https://www.gnu.org/licenses/>.

*/
package settings

import (
	"../build"
	"../locale"
	"../logger"
	"encoding/json"
	"flag"
	"fmt"
	"html/template"
	"io/ioutil"
	"os"
	"path/filepath"
)

// ## CONFIG CONTENT ##
//Used to read a configuration file
type ConfigType struct {
	HttpServer      HttpServerType
	Logging         LoggingType
	ShowHiddenFiles bool
	HumanFileSize   bool
	ModTimeFormat   string
}

type HttpServerType struct {
	Listen    string
	EnableTLS bool
	CertFile  string
	KeyFile   string
}

type LoggingType struct {
	Level         string // Info | Warning | Error
	LogRequest    bool
	Log404Request bool
}

//Default configuration
var ConfigDefault = ConfigType{
	HttpServer: HttpServerType{
		Listen:    ":8080",
		EnableTLS: false,
		CertFile:  "",
		KeyFile:   "",
	},
	Logging: LoggingType{
		Level:         "Info",
		LogRequest:    true,
		Log404Request: true,
	},
	ShowHiddenFiles: false,
	HumanFileSize:   true,
	ModTimeFormat:   "2006 Jan 2 15:04",
}

//Get the path to the file. Can return an empty string.
func GetFilePath(fileName string) string {
	if ArgConfigDir != "" {
		path := filepath.Join(ArgConfigDir, fileName)

		_, err := os.Stat(path)
		if err == nil {
			logger.Info.Println(locale.Load_file + ": " + path)
			return path
		}
		//logger.Info.Println(err)
		return ""
	}

	//Checking the USER path for the location of the config file
	for true {
		//Getting HOME environment variable
		variableHOME := os.Getenv(build.UserHomeEnvVar)
		if variableHOME == "" {
			logger.Warning.Println(build.UserHomeEnvVar + " " + locale.Env_var_is_empty)
			break
		}

		//Full path to file
		userFullPath := filepath.Join(variableHOME, build.UserConfigDir, fileName)

		//Checking the existence of a file
		_, err := os.Stat(userFullPath)
		if err == nil {
			return userFullPath
		}
		break
	}

	//Checking the SYSTEM path of the config file
	sysFullPath := filepath.Join(build.SysConfigDir, fileName)
	_, err := os.Stat(sysFullPath)
	if err == nil {
		return sysFullPath
	}

	//If both paths are wrong it returns an empty string
	return ""
}

//Read a configuration file
func ReadConfig(configFilePath string) ConfigType {
	//Getting the default configuration
	config := ConfigDefault

	//If an empty string is received, return the default value
	if configFilePath == "" {
		return config
	}

	//Reading a configuration file
	configFile, err := os.Open(configFilePath)
	if err != nil {
		logger.Warning.Println(err)
		return config
	}

	//Decoding a configuration file
	jsonParser := json.NewDecoder(configFile)
	err = jsonParser.Decode(&config)
	if err != nil {
		logger.Warning.Println(err)
		return config
	}

	return config
}

//Loads an HTML page template into RAM
func GetHtmlPageTemplate(indexHtmlFilePath string) *template.Template {
	//Custom template found
	if indexHtmlFilePath != "" {

		//Reading the html page template
		htmlPageTemplate, err := template.ParseFiles(indexHtmlFilePath)
		if err != nil {
			logger.Error.Fatal(err)
		} else {
			return htmlPageTemplate
		}
	}

	//The custom template is missing
	htmlPageTemplate := template.New("")

	htmlPageTemplate.Parse(templateDefault)

	return htmlPageTemplate
}

//Print Help
func PrintHelp() {
	fmt.Println(locale.Usage_flag)
	fmt.Println(`
-dir        ` + locale.Dir_flag + `
-config-dir ` + locale.Config_dir_flag + `
-listen     ` + locale.Listen_flag + `
-no-colors  ` + locale.No_colors_flag + `

-info       ` + locale.Info_flag + `
-version    ` + locale.Version_flag + `
-help       ` + locale.Help_flag)
}

//Global vars
var ArgDir string
var ArgConfigDir string
var ArgNoColors bool
var Config ConfigType
var HtmlTemplate *template.Template
var ResourcesDir string

func Main() {
	//Logging
	logger.SetLevel(Config.Logging.Level)

	//Flags
	flagSet := flag.NewFlagSet(build.Name, flag.ContinueOnError)
	flagSet.SetOutput(ioutil.Discard)

	ArgDir = *flagSet.String("dir", ".", locale.Dir_flag)
	ArgConfigDir = *flagSet.String("config-dir", "", locale.Config_dir_flag)
	argListen := flagSet.String("listen", "", locale.Listen_flag)
	ArgNoColors = *flagSet.Bool("no-colors", false, locale.No_colors_flag)
	argInfo := flagSet.Bool("info", false, locale.Info_flag)
	argVersion := flagSet.Bool("version", false, locale.Version_flag)
	argHelp := flagSet.Bool("help", false, locale.Help_flag)

	err := flagSet.Parse(os.Args[1:])
	if err != nil {
		fmt.Println(err)
		fmt.Println("")
		PrintHelp()
		os.Exit(2)
	}

	//Console setup
	if ArgNoColors != true {
		logger.UseColors()
	}

	//Displaying the build information
	if *argInfo == true {
		fmt.Println("Name:", build.Name)
		fmt.Println("Version:", build.Version)
		fmt.Println("SysConfigDir:", build.SysConfigDir)
		fmt.Println("UserHomeEnvVar:", build.UserHomeEnvVar)
		fmt.Println("UserConfigDir:", build.UserConfigDir)
		fmt.Println("LangEnvVar:", build.LangEnvVar)
		os.Exit(0)
	}

	//Displaying the version
	if *argVersion == true {
		fmt.Println(build.Name, build.Version)
		os.Exit(0)
	}

	//Displaying the help
	if *argHelp == true {
		PrintHelp()
		os.Exit(0)
	}

	//Reading a configuration file
	Config = ReadConfig(GetFilePath("config.json"))

	//HTML template content
	HtmlTemplate = GetHtmlPageTemplate(GetFilePath("index.tmpl"))

	//Path to the folder with resources
	ResourcesDir = GetFilePath("resources")

	//Rewriting config parameters
	if *argListen != "" {
		Config.HttpServer.Listen = *argListen
	}
}
